# Multiple Linear Regression

## ANTICIPATED TIME


2 hours

## BEFORE YOU BEGIN

[Simple Linear Regression](Simple_Linear_Regression.ipynb)

## WHAT YOU WILL LEARN


- What is the multiple regression model?
- What are interaction effects?
- What is a feature selection?
- What are diagnostics?
- How to use a pair plot?
- How to evaluate a model’s performance?
- How to work with categorical variables in regression?


## DEFINITIONS YOU’LL NEED TO KNOW

- Regression - model used to predict a number
- Multiple regression - a model that builds on the simple linear regression model by adding more predictors.
- Predictors - a variable that is used to predict the values of a dependent/response variable
- Response - the number you are trying to predict
- Multicollinearity - when two variables are linearly/closely related and have a strong connection.
- Interaction effects -  when the effect of one predictor variable depends on another predictor variable.
- Two-way interaction - how two variables work with each other.
- Dummy variables - when we assign numbers to categorical variables for analysis.
- Trendline - a line on a scatter plot that shows the general direction of the relationship between the two variables.


## SCENARIO:


Rin is a new member of the afterschool data club. She was observing the work Ethan, Diego, Angelina, and Kiana as they tried to make different predictions. Before they were using one variable to predict another one. Rin knows they have a lot of data, so she is wondering if multiple variables could be used to make a prediction for the variable she’s interested in.

Applying Multiple Linear Regression will help the club make better predictions considering multiple data points. In this lesson, we’re going to understand how this all works!  


## WHAT DO I NEED TO KNOW?



In the [Simple Linear Regression notebook](Simple_Linear_Regression.ipynb), we learned about linear regression where we used one variable to predict another one. We learned how ordinary least squares and the best-fit line can help us with this. But we know life is kinda messy and we often see a lot of variables that are important.

Let’s think about trying to predict someone’s number grade on a test, you can think about things like….

…attendance?
…time spent studying?
…class participation?
…and grades from previous assignments?

Simple linear regression uses one variable, but multiple regression can help us figure out which of many variables is most important for receiving a good score on the test. Instead of looking at one variable, we look at (you guessed it) multiple variables to understand what areas need improvement to achieve a higher grade!

**What variables should I put in my model?**

But wait!  **Multicollinearity** is when two variables are linearly/closely related and have a strong connection. This means that some variables are measuring the same thing, and we may not need certain variables. So it makes it really hard for the model to pinpoint the impact of the variable. It’s important to decide which variables should be used because if variables have a strong connection, we won’t be able to get an estimate for the model’s parameters.


**Interaction Effects - How All the Predictor Variables Impact Each Other**

Remember, multiple linear regression adds the effects of predictors together. Predictors don't influence each other, they only affect the outcome variable.

But what if one predictor variable influences another predictor variable? We can model this with a new idea called an interaction. Interactions don't add the different variables, instead, they are multiplied by each other. An interaction multiplies two (or more) variables together to make a new predictor. When we multiply them, the result then becomes a predictor that is added to other predictors in the normal way.

So how does this work?

Let’s say we discover older (Age) students with high attendance can predict higher grades.  This might mean that the age of the student and their attendance can especially affect grades, making it an interaction effect. You might think of it as

Age * Attendance = Grades

Occasionally, when two variables are looked at we can see that they work well together, which makes this a **two-way interaction**. There can also be three-way interactions (grades, attendance, class participation), although they are not as common.

**What if I have Categories instead of Numbers?**

When we think about our example of predicting a student’s final grade, there are variables that we need to measure that aren’t numerical such as the student’s overall attitude towards attending school. When we measure things that aren’t numbers, we call them **categorical variables**, as we discussed in the [Data Science and Nature of Variables notebook](Datascience_and_the_Nature_of_Data.ipynb). Categorical variables represent categories or groups. For example, we can ask students how they feel about school based on the categories “positive” or “negative.”
Even though these aren’t number responses, to use these variables in regression analysis, we must change them into numbers. To do this, we need to use **dummy variables**. To use dummy variables, we will assign each category a 0 or 1. For example, “negative” can be 0, and “positive” can be 1. Now that we’ve transformed numbers to categories, we can use numbers to analyze categorical variables and understand how these categories have an impact on the outcome.
How does this work in the model? When the category is "negative" and the value is zero, then the predictor goes to zero (anything multiplied by zero is zero).


## YOUR TURN



### Goal 1: Importing the Pandas Library

Need extra tools to help solve this problem? Well, we can bring in extra ‘libraries’ to help us do extra data science stuff. You can think of it as an ‘add-on’. In this case, we bring in pandas, which is a popular library for doing data science stuff.  

#### Blockly

**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.

Bring in the IMPORT menu, which can be helpful to bring in other data tools. In this case, we're bringing in the **import** block.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **pandas**, which will bring in some cool data manipulation features.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **import** and **package** together in a single variable. This handy feature helps cut down on all the typing later on. You can call it whatever is easiest for you to remember. In the example below, we’ve put everything into **pd**, and we type it in the open area.



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GcxNkjYXkAAlpKD?format=png&name=240x240)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.


**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **pandas**, which will bring in some cool data manipulation features.


**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the ‘import’ and ‘package’ together in a single variable. This handy feature helps cut down on all the typing later on. Feel free to use whatever name you want that will help you remember it later on. In the example below, we’ve put everything into **pd**.


**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	

![](https://pbs.twimg.com/media/GZmkVCYWEA4oGso?format=jpg&name=small)

</details>

**Your Turn**: Now it’s your turn! We’re going to dive into the pandas package, which helps us with some really cool data science things. First, let’s import the package and assign it to the variable “pd” to make it easier to use throughout our notebook.


In [3]:
#freehand code 


**Explanation**: *Congrats!  Your attempts finally made it!  Now you have successfully imported the "pandas" package as the variable "pd"*.

### Goal 2: Load the training dataset

Let’s bring in the data from a CSV file to visualize the data for further analysis and manipulation.

#### Blockly


**Step 1 - Write out the variable name you want to use**

Now that we’re all set with our new package to help us do cool things, let’s bring the data into a variable and call it **train**.

In Blockly, bring in the VARIABLES menu.



**Step 2 - Assign the dataframe to the variable you created**

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.

In Blockly, go to the Variables and drag the Set block for the **train** variable. This will allow us to assign the result of a function call to the variable. A function is basically code that does a specific task for us.



**Step 3 - Bring in the data**

Now we need to look at the file that has all our data. To load our dataframe, we’ll use a simple command to bring in the file we need (CSV….Comma Separated Values). Let’s say we have a file called ‘AirQualityCars.csv' in the folder **‘datasets’**. We’re telling Python to read the CSV file and store it in a variable called **train**.

From the Variable menu, drag a DO block using the **pd** variable, and go ahead with the do operation **read_csv**. The read_csv function reads a CSV file and returns a DataFrame object.

In our case, let’s bring in the "datasets/AirQualityCars.csv" (use the Quotes from the TEXT menu) because that is what Angelina is working with.



**Step 4 - Display the variable**

Let’s see it now by ‘displaying’ and showing our work.

Drag the **train** variable to the workspace, making it available for further use in our program. This step is more of a visualization step, as it allows us to see the variable in the Blockly workspace.



**Step 5 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab5Z-TXsAAQAWr?format=jpg&name=small)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Write out the variable name you want to use**

Now that we’re all set with our new package to help us do cool things, let’s bring the data into a variable called train. Think of it as a digital spreadsheet with much more power to analyze and manipulate the data!

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.



**Step 2 - Bring in the data**

Now we need to look at the file that has all our data.

To load our dataframe, we’ll use a simple command to bring in the file we need (CSV….Comma Separated Values). Let’s say we have a file called ‘AirQualityCars.csv' in the folder **‘datasets’**. We’re telling Python to read the CSV file and store it in a variable called **train**. For this function, we need to specify the code as “pd.read_csv”, which makes the code read the csv file. This variable is now our dataframe!

In our case, let’s bring in the “datasets/AirQualityCars.csv" (user the Quotes from the TEXT menu) because that is what our group is working with



**Step 3 - Assign the train to the variable you created**

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.



**Step 4 - Print the variable**

Let’s see it now by ‘printing’ and showing our work.



**Step 5 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!
<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab5c61WwAApSiM?format=png&name=small)

</details>


**Your Turn**: Now it’s your turn!  Let’s dive in and start working with the data!  We’ll begin by loading it into a dataframe, which will allow us to easily interact with and analyze the dataset.


In [4]:
#freehand code 


Brand       Model  Volume  Weight  CO2Emission
0          Toyota     Corolla    12.3    1300          120
1          Toyota       Camry    14.0    1495          135
2          Toyota        RAV4    15.0    1680          140
3          Toyota       Prius    12.5    1375           90
4          Toyota  Highlander    16.5    1965          165
5            Ford       F-150    19.5    2200          255
6            Ford     Mustang    13.0    1655          180
7            Ford    Explorer    16.8    2020          200
8            Ford       Focus    12.0    1350          115
9            Ford      Escape    15.4    1580          150
10      Chevrolet   Silverado    19.8    2260          260
11      Chevrolet      Malibu    13.5    1460          135
12      Chevrolet     Equinox    15.3    1705          155
13      Chevrolet      Camaro    13.2    1685          190
14      Chevrolet    Traverse    17.5    2030          210
15          Honda       Civic    12.2    1290          110
16          Honda      Accord    13.6    1490          130
17          Honda        CR-V    14.8    1610          140
18          Honda        HR-V    14.0    1425          115
19          Honda       Pilot    16.7    1975          180
20         Nissan      Altima    13.1    1470          130
21         Nissan       Rogue    15.0    1640          145
22         Nissan      Sentra    12.0    1250          110
23         Nissan    Frontier    18.0    2000          195
24         Nissan      Maxima    13.9    1600          145
25            BMW    3 Series    13.0    1570          140
26            BMW          X5    16.5    2190          195
27            BMW    5 Series    14.0    1825          160
28            BMW          X3    15.0    1820          170
29            BMW    7 Series    15.5    2195          205
30  Mercedes-Benz     C-Class    13.1    1530          130
31  Mercedes-Benz     E-Class    14.8    1760          150
32  Mercedes-Benz         GLC    15.9    1835          170
33  Mercedes-Benz         GLE    16.5    2130          190
34  Mercedes-Benz     S-Class    15.8    2080          175
35           Audi          A4    13.2    1520          135
36           Audi          Q5    15.3    1765          160
37           Audi          A6    14.0    1695          145
38           Audi          Q7    16.5    2170          200
39           Audi          A3    12.5    1370          120
40     Volkswagen        Golf    12.0    1320          115
41     Volkswagen      Tiguan    15.4    1650          140
42     Volkswagen      Passat    13.5    1470          125
43     Volkswagen       Jetta    13.0    1420          120
44     Volkswagen       Atlas    17.0    2000          190
45        Hyundai     Elantra    12.0    1275          115
46        Hyundai      Tucson    15.0    1590          140
47        Hyundai    Santa Fe    16.0    1810          165
48        Hyundai        Kona    13.1    1410          125
49        Hyundai      Sonata    13.8    1440          130

**Explanation**:  *The training dataset contains information about different car models and their CO2 emissions. Each row includes the brand (e.g., Toyota, Ford), model, volume (likely engine size), weight, and CO2Emission level. Heavier cars or those with larger engine volumes tend to have higher CO2 emissions because they use more fuel. For example, smaller cars like the Toyota Prius have lower emissions, while larger vehicles like the Ford F-150 have higher emissions. This data helps a model learn the relationship between features like volume and weight and their impact on CO2 emissions*.

### Goal 3: Import the seaborn library

Before, we’ve been using Plotly to help with the visualization, but we’ll switch it up a bit and use the Seaborn library which has some other visualizations for the correlations we’ll be looking at.

#### Blockly

**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.

Bring in the IMPORT menu, which can be helpful to bring in other data tools. In this case, we're bringing in the **import** block.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **seaborn**, which will bring in some cool data manipulation features.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **import** and **seaborn** together in a single variable. This handy feature helps cut down on all the typing later on. You can call it whatever is easiest for you to remember. In the example below, we’ve put everything into **sns**, and we type it in the open area.



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	

![](https://pbs.twimg.com/media/Gbdo5AabUAA8svq?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.


**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **seaborn**, which will bring in some cool data manipulation features.


**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the ‘import’ and ‘package’ together in a single variable. This handy feature helps cut down on all the typing later on. Feel free to use whatever name you want that will help you remember it later on. In the example below, we’ve put everything into **sns**.


**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the data science magic!

<details>
    <summary>Click to see the answer...</summary>

 
![](https://pbs.twimg.com/media/GbfSdtqaQAA2hpQ?format=png&name=small)

</details>

**Your Turn**: Now it’s your turn! We’re going to dive into the seaborn package, which helps us with some really cool data science things. First, let’s import the package and assign it to the variable “sns” to make it easier to use throughout our notebook.


In [5]:
#freehand code 


**Explanation**: *Now you have a new graphic library, seaborn, which is a tool that makes it easy to create nice-looking charts and graphs in Python*.

### Goal 4: Analyze correlations with pair plot

Before, we used a scatter plot that helped show the relationship with 1 variable. Let’s use a pair plot to find the relationships among multiple variables, which will help us with our multiple linear regression

#### Blockly


**Step 1 - Create a pair plot**

Let’s see if we can explore the correlations between the variables. Let’s start with a visualization called ‘pairplot’ that helps visualize the relationship among all the variables used.

From the Variables menu, get a DO block for the **sns** variable. With it, select the operation pairplot from the Seaborn library.



**Step 2 - Add the parameters to the graph**

Inside the pairplot() method, we add our parameter for the plot. For the pairplot, we add our parameter(s). In this case, we’ll bring in **train** variable that has the dataframe and the variables we’ll look at.

Drag the **train** variable from the variable menu to connect to it. The pair plot will show histograms for each variable on its diagonal and, on each one of the grids, a scatter plot between variables.



**Step 3 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbdrM54aoAAKF14?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Present a pairplot**

Let’s see if we can explore the correlations between the variables. Let’s start with a visualization called ‘pairplot’ that helps visualize the relationship among all the variables using.

For the pairplot, we call the pairplot() method from the seaborn library.

`sns.pairplot()`



**Step 2 - Add the parameters of the graph**

Inside the pairplot() method, we add our parameter for the plot. For the pairplot, we add our parameter(s). In this case, we’ll bring in **train** variable that has the the dataframe and the variables we’ll look at.

`sns.pairplot(train)`



**Step 3 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!
<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbfSDsIaIAARGoc?format=png&name=small)

<details>
    <summary>Click to see the answer...</summary>
	

**Your Turn**: Now it’s your turn! Use pair plot to analyze the relationships between variables.

In [1]:
#freehand code 


**Explanation**: *The pairplot shows histograms for each variable on its diagonal, and on each of the grids a scatter plot between variables. We can see that as volume and weight increase, CO2emission also tends to increase. The plots with upward trends (like volume vs. CO2 and weight vs. CO2) indicate that cars with larger engines and more weight generally emit more CO2. This pattern suggests that both engine size and weight play a role in how much CO2 a car produces, which can hint that there is a correlation between the features and the label, so we can create a linear model to predict it*.

### Goal 5: Calculate correlations

Now that we’ve brought in our dataset loaded and the right library to help us, let’s see if we can find any relationships between the variables we can calculate.

#### Blockly

**Step 1 - Call the corr() method of the dataframe**

To get the correlation of variables, we first need to call the corr() method to help calculate the correlations. We’ll apply it to the data we have in our **train** variable.

From the VARIABLES menu, WITH the **train** variable DO the **corr** function.



**Step 2 - Run correlations for only numeric values**

We tell our code **numeric_only=True** (FREESTYLE menu), which means we only want to consider numeric columns in the calculation.



**Step 3 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbeNDp7aAAAOsGC?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Call the corr() method of the dataframe**

To get the correlation of variables, we first need to call the **corr**() method to help calculate the correlations. We’ll apply it to the data we have in our **train** variable.

`train.corr()`



**Step 2 - Run correlations for only numeric values**

We tell our code numeric_only=True (FREESTYLE menu), which means we only want to consider numeric columns in the calculation.

`train.corr(numeric_only=True)`



**Step 3 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbfR10JbEAAv_3K?format=png&name=small)

</details>


**Your Turn**: Now it’s your turn to calculate correlations!

In [7]:
#freehand code 


Volume    Weight  CO2Emission
Volume       1.000000  0.897033     0.876196
Weight       0.897033  1.000000     0.922855
CO2Emission  0.876196  0.922855     1.000000

**Explanation**: *The table shows the correlation values between volume, weight, and CO2 emissions. A correlation value close to 1 means a strong positive relationship, whereas one-factor increases, the other tends to increase as well. Here, volume and CO2 emissions have a correlation of 0.876, meaning that cars with larger volumes tend to emit more CO2. Weight and CO2 emissions have an even stronger correlation of 0.923, showing that heavier cars are likely to produce more CO2. This data confirms that both volume and weight are closely related to CO2 emissions, with weight having a slightly stronger influence*.

### Goal 6: Import the Plotly.Express Library

We’ve already brought pandas to help with data science. Let’s bring in Plotly Express to help with some fancy-pants visualizations.

#### Blockly


**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.

Bring in the IMPORT menu, which can be helpful to bring in other data tools. In this case, we're bringing in the **import** block.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **plotly.express**, which will bring in some cool data manipulation features.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **import** and **package** together in a single variable. This handy feature helps cut down on all the typing later on. You can call it whatever is easiest for you to remember. In the example below, we’ve put everything into **px**, and we type it in the open area.


**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbeNeZYaAAAc6yK?format=png&name=small)

</details>


In [ ]:
#blocks code


#### Freehand


**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.


**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **plotly.express**, which will bring in some cool data manipulation features.


**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **‘import’** and **‘package’** together in a single variable. This handy feature helps cut down on all the typing later on. Feel free to use whatever name you want that will help you remember it later on. In the example below, we’ve put everything into **px**.


**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbfSbjibcAEK55L?format=png&name=360x360)

</details>


**Your Turn**: Now it’s your turn! Import plotly express to get some great visuals!


In [8]:
#freehand code 


**Explanation**: *The line import plotly.express as px imports the Plotly Express library, a high-level interface for creating interactive and dynamic visualizations in Python*.

### Goal 7: Present a scatter plot

Scatter plots help us to look at each data point when it comes to interval ratio data. The scatter plot shows us the relationship between two variables in a data set. One variable is plotted on the X-axis, while the other variable is plotted on the Y-axis. They are super handy for finding the relationship between different numeric variables.

#### Blockly


**Step 1 - Call the scatter function from plotly**

Let’s see if we can explore the correlations between the two variables. Let’s start with a visualization that helps ‘scatter’ all the data in one visualization.

From the Variables menu, get a DO block for the **px** block. With that, select the **scatter** operation. That will generate the scatter plot.



**Step 2 -  Saying what data to use for the scatter plot**

So what are we going to look at? Simple, you just have to inform the x-axis and y-axis variables.

Inside the scatter() method, we add our parameters for the plot. First, we have the **train**,  which is the data containing the data to plot.  From the Variables menu, get the **train** variable that will contain the data that you want to generate the scatter plot with. Then, get two Quote “” blocks from the Text menu.



**Step 3 -  Tell plotly what columns to put on the axis**

Now we need to inform the x-axis and y-axis variables.  Type for each of the names of the columns of the x-axis and y-axis as **‘Weight’** and '**CO2Emission**'.     



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbE_1liWwAAaKro?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Call the scatter function from plotly**

Let’s see if we can explore the correlations between the two variables. Let’s start with a visualization that helps ‘scatter’ all the data in one visualization.

To make the scatterplot, we first call the **scatter**() function from the px library.

`px.scatter()`



**Step 2 - Saying what data to use for the scatter plot**

So what are we going to look at? Simple, you just have to inform the x-axis and y-axis variables.

Inside the scatter() method, we add our parameters for the plot. First, we have the **train**, which is the data containing the data to plot



**Step 3 - Tell plotly what columns to put on the axis**

‘Weight’ is the column to be plotted on the x-axis, representing the weight of the cars. '**CO2Emission’** is the amount of CO2 emissions we have recorded. This second thing is the column to be plotted on the y-axis.

Inside the scatter() method, we add the following parameters to make our scatterplot.
- **train**: the name of the dataframe
- **Weight**: the data to plot in the X-Axis
- **CO2Emission**: the data to plot in the Y-Axis



**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbE_xP_WgAA5J88?format=png&name=small)

</details>

**Your Turn**: Now you give a go! Let’s start by creating a scatterplot using px.scatter so we can better visualize the relationship between each of our variables.


In [1]:
#freehand code 


**Explanation**: *The scatter plot shows how the weight of an object (like a car) is related to the amount of CO2 it emits. Each dot represents a sample with its specific weight and CO2 emission level. We can see a general upward trend: as the weight increases, CO2 emissions tend to be higher as well. This suggests that heavier objects or vehicles usually produce more CO2. However, the points don't form a perfect line, which means there are other factors besides weight that also affect CO2 emissions*.

### Goal 8: Create a scatter plot and trendline to visualize the correlation between two variables

We’re going to create a scatter plot to explore the correlation between the weight of cars and their CO2 emissions. You'll be able to visualize the data and see the trendline that shows their relationship. Let's get started!

#### Blockly


**Step 1 - Call the scatter function from plotly**

Let’s see if we can explore the correlations between the two variables. Let’s start with a visualization that helps ‘scatter’ all the data on one visualization.

From the Variables menu, get a DO block for the **px** block. With that, select the **scatter** operation. That will generate the scatter plot.



**Step 2 -  Saying what data to use for the scatter plot**

So what are we going to look at? Simple, you just have to inform the x-axis and y-axis variables.

Inside the scatter() method, we add our parameters for the plot. First, we have the **train**,  which is the data containing the data to plot.  From the Variables menu, get the **train** variable that will contain the data that you want to generate the scatter plot with. Then, get two Quote “” blocks from the Text menu.



**Step 3 -  Tell plotly what columns to put on the axis**

Now we need to inform the x-axis and y-axis variables.  Type for each of the names of the columns of the x-axis and y-axis as **‘Weight’** and '**CO2Emission**'.     



**Step 4 - Show the linear regression**

Lastly, we want to add a trendline to our scatter plot to help visualize the relationship between the x and y variables.

In this case, using **freestyle**, type **trendline='ols'**. This will be the least ordered square that we talked about earlier. This will give us the scatterplot, which will show the relationship between the weight of the car and the amount of CO2 it emits.



**Step 5 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab603uXMAA2kms?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Call the scatter function from plotly**

Let’s see if we can explore the correlations between the two variables. Let’s start with a visualization that helps ‘scatter’ all the data in one visualization.

To make the scatterplot, we first call the scatter() function from the px library.

`px.scatter()`



**Step 2 - Saying what data to use for the scatter plot**

So what are we going to look at? Simple, you just have to inform the x-axis and y-axis variables.

Inside the scatter() method, we add our parameters for the plot. First, we have the **train**, which is the data containing the data to plot



Step 3 - Tell plotly what columns to put on the axis

**‘Weight’** is the column to be plotted on the x-axis, representing the weight of the cars. '**CO2Emission’** is the amount of CO2 emissions we have recorded. This second thing is the column to be plotted on the y-axis.

Inside the scatter() method, we add the following parameters to make our scatterplot.
- **train**: the name of the dataframe
- **Weight**: the data to plot in the X-Axis
- **CO2Emission**: the data to plot in the Y-Axis



**Step 4 - Show the linear regression**

Lastly, we want to add a trendline to our scatter plot to help visualize the relationship between the x and y variables. In this case, we’ll type type **trendline='ols'**. This will be the least ordered square that we talked about earlier
trendline=’ols’: this parameter adds a trendline to the scatter plot to help visualize the relationship between the x and y variables

This will give us the scatterplot, which will show the relationship between the weight of the car and the amount of CO2 it emits.

`px.scatter(train,'Weight','CO2Emission',trendline='ols')`




**Step 5 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab6yUOXcAIM6_H?format=jpg&name=small)

</details>

**Your Turn**: Test it out yourself! Firstly we will create a scatterplot to explore our two variables: car weight and CO2 emissions, to see their relationship. Make sure to set the x and y axis with the correct data then we can add a trendline to see the pattern between the variables and better understand their relationship!


In [2]:
#freehand code 


**Explanation**: *The scatterplot shows the relationship between the weight of an object (like a car) and the amount of CO2 it emits. The dots represent data points for different weights and their corresponding CO2 emissions. As we can see, there’s a trend where heavier objects generally produce more CO2 emissions, shown by the upward slope of the trend line. The trend line helps summarize the overall direction of the data. In simple terms, this means that, typically, the heavier something is, the more CO2 it tends to emit more gas*.


### Goal 9: Import the linear model library

Let’s bring in a library/package to help with the linear regression that will help us with our analysis and other data science tasks.

#### Blockly



**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.

Bring in the IMPORT menu, which can be helpful to bring in other data tools. In this case, we're bringing in the **import** block.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **sklearn.linear_model**, which will bring in some cool data manipulation features.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **import** and **package** together in a single variable. This handy feature helps cut down on all the typing later on. You can call it whatever is easiest for you to remember. In the example below, we’ve put everything into **lm**, and we type it in the open area.



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbFDOz7XoAAVLtr?format=png&name=small)

</details>


In [ ]:
#blocks code


#### Freehand


**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.


**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **sklearn.linear_model**, which will bring in some cool data manipulation features.


**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the ‘import’ and ‘package’ together in a single variable. This handy feature helps cut down on all the typing later on. Feel free to use whatever name you want that will help you remember it later on. In the example below, we’ve put everything into **lm**.


**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbFDMCWW8AABSiy?format=png&name=small)

</details>



**Your Turn**: Now it’s your turn! We’re going to dive into the sklearn.linear_model package, which helps us with some really cool data science things. First, let’s import the package and assign it to the variable “lm” to make it easier to use throughout our notebook.


In [3]:
#freehand code 


**Explanation**: *With the scikit-learn library, specifically the linear model module, which includes methods for creating linear regression models*.



### Goal 10: Setting up our linear model

Let’s create a model to help with the training that we will do for our dataset.  

#### Blockly

**Step 1 - Create and assign a variable**

Now that we’re all set with our new package to help us to do cool things, let’s bring the data into a variable and call it **regr**. Think of it as a digital spreadsheet with much more power to analyze and manipulate the data!

In Blockly, bring in the VARIABLES menu. On the "Variables" menu, click Create Variable, type a name for our model, **regr**. Then, drag a "SET" block to the workspace for the created variable. This block allows us to create a new variable and assign a value to it.



**Step 2 - Create the linear regression model**

Using the neighbors library, we call the LinearRegressor() to create the linear regression model.

From the Variable menu, drag a Create block for the lm variable. On the create list box select the option **LinearRegressor**. This specifies the type (class) of object we want to create, which is the KNeighborsRegressor from the neighbors module.

Get a Create block for the lm variable from the Variables menu. With that, a new object of the model, **LinearRegressor**, is created. The **LinearRegressor** is a type of regression model that uses one specific number to predict one other number,



**Step 3 - Store the linear regressor model in a variable**

We can now connect the **regr** variable with the **LinearRegressor** model.



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbFD3KsXwAAqIWJ?format=png&name=small)

</details>


In [ ]:
#blocks code


#### Freehand


**Step 1 - Create the linear regression model**

Using the linear model library, we call the LinearRegression() method

`lm.LinearRegression()`



**Step 2  - Store the regression model in a variable**

Now that we’re all set with our new package to help us to do cool things, let’s bring the data into a variable called **regr**. Think of it as a digital spreadsheet with much more power to analyze and manipulate the data!

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.

`regr = lm.LinearRegression()`





**Step 3 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbFD57vWsAAOizJ?format=png&name=small)

</details>


**Your Turn**: Now give it a go and see what you get!


In [4]:
#freehand code 


**Explanation**:  *You have created a linear regression model. By creating this model, we're preparing to analyze how one variable, like weight, affects another variable, like CO2 emissions, by finding the line that best fits the data points. Once trained, this model can help predict CO2 emissions based on weight or any other factor it’s trained on*.

### Goal 11: Train and Score the Regressor Model

Now that we’ve brought in our linear regression model, let’s train the model to see how it will learn from the data points that we have in the file.

#### Blockly


**Step 1 - Prepare to train the model**

To train data using the classifier model, we use the model and call the fit() method from it. This will use the fit method to train the model that we want to train on.  

From the Variable menu, drag the DO block for the **regr** variable, and select the **fit** method as the do operation. This specifies the function we want to call, which is the fit method of the linear regression.



**Step 2 - Have the training features ready**

The next step for training the model is to select the features to train the regressor. In this step, we select the features and add them as a dataframe in the parameter. In this case, the model will train (learn) the linear regressor variables and use it to predict our other variable

From the Lists menu, drag a dictVariable, and select the **train** variable from the list of available variables. Also, from the Lists menu, you will get a Create List block. Using the Gear icon, add up to 2 items. For each one of the items, add a Text (a Quote “” from the Text menu), as follow:  "**Weight**" and **“Volume”**



**Step 3 -  Have the training label ready**

So what are we trying to predict? Next, we need to add the data labels for the selected features. We add the data labels (**CO2Emission**) as a parameter in the fit() method.

From the Lists menu, drag a dictVariable, and select the "train" variable from the list of available variables. From the Text menu, get a Quote “” block and add a Text **CO2Emission**. This is the target value applied to train (fit) the model.



**Step 4 - Measure the correctness on the training dataset**

To measure the correctness of the model, we will use the score method() from the neighbors library. Just as in the previous step, we will just replace the fit() method with the score() method. Based on the ‘fit’, we will try to see how much we were able to predict in our training dataset.

This will give us the linear regression correctness score. A good score will be closer to 1 (ie - 100). Medium might be more like .95 (95% accurate). Not great would be .90 (90%). It depends on the topic you are looking at.

Right-click on the "**regr.fit**" block and select "Duplicate" from the context menu. This creates a copy of the block. Within the duplicated block, click on the method dropdown menu and select "**score**" from the list of available methods. The score method will work similarly to fit, and will use the training features and label to measure how much of the training data was learned.  



**Step 5 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbeQcPJbYAADVoO?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Prepare to train the model**

To train data using the regression model, we use the model and call the fit() method from it. This will use the ‘fit’ method to train the model that we want to train on.  
  
`regr.fit()`



**Step 2 - Have the training features ready**

The next step for training the model is to select the features to train the classifier. In this step , we select the features and add them as a dataframe in the parameter. In this case, we select the **‘Weight’** and **‘Volume’** feature as our training feature.

`regr.fit(train[['Weight',’Volume’]])`




**Step 3 - Have the training label ready**

So what are we trying predict? Next, we need to add the data labels for the selected features. We add the data label (**CO2Emission**) as a parameter in the fit() method.

`regr.fit(train[['Weight',’Volume’]],train['CO2Emission'])`



**Step 4 - Measure the correctness on the training dataset**

To measure the correctness of the model, we will use the score method() from the SKLearn.linear_model library. Just as the previous step, we will just replace the fit() method with the **score**() method. Based on the ‘fit’, we will try to see how much we were able to predict in our training dataset.

This will give us the linear regression correctness score. A good score will be closer to 1 (ie - 100). Medium might be more like .95 (95% accurate). Not great would be .90 (90%). It depends on the topic you are looking at.

`regr.score(train[['Weight',’Volume’]],train['CO2Emission'])`



**Step 5 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	

![](https://pbs.twimg.com/media/GbfVH3TaEAAWl69?format=jpg&name=small)

</details>

**Your Turn**: Now it’s your turn to train the model!


In [5]:
#freehand code 


**Explanation**: *You have trained the model using data from the train dataset. It takes weight as the input feature and CO2 emission as the output we want to predict. The fit function makes the model learn the relationship between weight and volume, with CO2 emissions from this data. You have also calculated the model's R-squared score on the training data. The R-squared score, which is 0.86 (or 86%), shows how well the model’s predictions match the actual CO2 emissions. A score close to 1 (or 100%) means the model is doing a good job. Here, 0.86 means the model explains 86% of the variation in CO2 emissions based on weight and volume, indicating it's a fairly accurate model*.

### Goal 12: Load the testing dataset

So we’ve looked at the training dataset to learn something about our data. How about applying it to the rest of the dataset and seeing how good our predictions are?

#### Blockly


**Step 1 - Write out the variable name you want to use**

Now that we’re all set with our new package to help us do cool things, let’s bring the data into a variable and call it **test**. Think of it as a digital spreadsheet with much more power to analyze and manipulate the data!

In Blockly, bring in the VARIABLES menu.



**Step 2 - Assign the dataframe to the variable you created**

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.

In Blockly, go to the Variables and drag the Set block for the **test** variable. This will allow us to assign the result of a function call to the variable. A function is basically code that does a specific task for us.



**Step 3 - Bring in the data**

Now we need to look at the file that has all our data.

To load our dataframe, we’ll use a simple command to bring in the file we need (CSV….Comma Separated Values). Let’s say we have a file called ‘AirQualityCtest.csv' in the folder **‘datasets’**. We’re telling Python to read the CSV file and store it in a variable called **test**.

From the Variable menu, drag a DO block using the **pd** variable, and go ahead with the do operation **read_csv**. The read_csv function reads a CSV file and returns a DataFrame object.

In our case, let’s bring in the “datasets/AirQualityCtest.csv" (use the Quotes from the TEXT menu) because that is what Angelina is working with.



**Step 4 - Display the Variable**

Let’s see it now by ‘displaying’ and showing our work.

Drag the **test** variable to the workspace, making it available for further use in our program. This step is more of a visualization step, as it allows us to see the variable in the Blockly workspace.



**Step 5 - Connect the blocks to run the code**

Connect the blocks and run the code!
<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab-trtWgAAOONP?format=jpg&name=small)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Write out the variable name you want to use**

Now that we’re all set with our new package to help us to do cool things, let’s bring the data into a variable called **test**. Think of it as a digital spreadsheet with much more power to analyze and manipulate the data!

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.



**Step 2 - Bring in the data**

Now we need to look at the file that has all our data.

To load our dataframe, we’ll use a simple command to bring in the file we need (CSV….Comma Separated Values). Let’s say we have a file called ‘AirQualityCtest.csv' in the folder **‘datasets’**. We’re telling Python to read the CSV file and store it in a variable called **test**. For this function, we need to specify the code as “pd.read_csv”, which makes the code read the csv file. This variable is now our dataframe!

In our case, let’s bring in the “datasets/AirQualityCtest.csv" (user the Quotes from the TEXT menu) because that is what our group is working with.



**Step 3 - Assign the dataframe to the variable you created**

Just like we did before, let’s type out a variable name. Rather than type out the full file name for our data, this easy to remember name will hold the data we bring in.



**Step 4 - Print the variable**

Let’s see it now by ‘printing’ and showing our work.



**Step 5 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/Gab-wAQXMAAIVs8?format=png&name=small)

</details>

**Your Turn**: Now it’s your turn!  Let’s dive in and start working with the data!  We’ll begin by loading it into a dataframe, which will allow us to easily interact with and analyze the dataset.


In [14]:
#freehand code 


Brand     Model  Volume  Weight  CO2Emission
0  Porsche       911      13    1570          153
1  Porsche   Cayenne      16    2000          199
2  Porsche     Macan      15    1855          152
3  Porsche  Panamera      15    1850          170
4  Porsche    Taycan      14    2290          245
5     Fiat       500      10     965           82
6     Fiat     Panda      11     950           79
7     Fiat      Tipo      12    1270          104
8     Fiat      500X      13    1420          126
9     Fiat     Doblo      14    1600          152

**Explanation**: *You have loaded the testing dataset, which is essential to see if our model makes accurate predictions on new, unseen data, ensuring it generalizes well beyond the training data. In this dataset, we have different Porsche models with their Weight and actual CO2Emission values. Testing the model on this data shows if it can predict CO2 emissions accurately based on features like weight, confirming the model's reliability*.

### Goal 13: Predict Labels for Testing Dataset (i.e., - predict the rest of the data)

So far, we’ve taken a smaller part of all our data to train and try and learn something about it. Can we take what we’ve learned from the training and use it to predict the rest of our dataset?

#### Blockly

**Step 1 - Write out the variable name you want to use**

Now that we’re all set with our new package to help us to do cool things, let’s bring the data into a variable and call it **predictions**.

From the Variables menu, click Create Variable, and type **predictions**. On the same menu, drag the Set block of the prediction variable. This variable will hold the result of the prediction.



**Step 2 - Prepare the predict operation**

So let’s take the regr variable from before and try to predict the label of the new dataset for **CO2Emission**. Let’s start by using the predict() method from the linear regression model.

From the Variables menu, get a DO block, for the **regr** variable. With that select the operation **predict**.



**Step 3 - Set the test features**

Inside the predict() method,we  provide the test features from the test data. This will use the 1 features (ie - columns) to predict the labels.

From the Lists menu, drag a dictVariable, select the "test" variable from the list of available variables. Also, from the Lists menu, you will get a Create List block. Using the Gear icon, add up to 2 items. For each one of the items, add a Text (a Quote “” from the Text menu), as follow: **Weight** and **Volume**. These are the feature names applied to predict the target label on the testing dataset. Store the output of the model prediction in the "**predictions**" variable. This variable will now hold the result of the prediction.



**Step 4 - Assign the predictions to the variable you created**

Next, we store the prediction labels into a variable ‘predictions’. To do that we have to connect the SET **predictions** variable to the r**egr.predict**() block.



**Step 5 - Display the predictions**

Let’s see it now by ‘printing’ and showing our work.

Drag the **predictions** variable to the workspace, making it available for further use in our program. This step is more of a visualization step, as it allows us to see the variable in the Blockly workspace.



**Step 6 - Connect the blocks to run the code**

Connect the blocks and run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbeSDKLbsAAT4EG?format=png&name=small)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Select the test features**

So let’s take the **regr** variable from before and try to predict the label of the new dataset. Let’s start by using the predict() method from the linear regression model.

`regr.predict()`



**Step 2 - Set the test features**

Inside the predict() method,we  provide the test features from the test data. This will use the features (ie - data we have in a column) to predict the labels - **Weight** and **Volume**.

`regr.predict(train[['Weight',’Volume’]])`



**Step 3 - Store the predictions in a variable**

Next, we store the predicted results into a variable ‘predictions’

`predictions = regr.predict(train[['Weight',’Volume’]])`


**Step 4 - Display the predictions**

Finally, we display the prediction labels using ‘predictions’

`predictions`



**Step 5 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
	
![](https://pbs.twimg.com/media/GbfQzORbcAATBir?format=png&name=small)

</details>

**Your Turn**: Now it’s your turn to get predictions from the model.


In [15]:
#freehand code 


**Explanation**: You have used the predict function to estimate results based on new input data, Weight and Volume, from the testing dataset. In this case, the code tells the model to make predictions using values from the Weight and Volume columns in the test dataset. These columns are the features of the model used to predict the outcome, CO2 emissions.

### Goal 14: Import the metrics library

Need extra tools to help solve this problem? Well, we can bring in extra ‘libraries’ to help us do extra data science stuff. You can think of it as an ‘add-on’. In this case, we bring in SKLearn, which is a popular library that helps us with the predictions.

#### Blockly

**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.

Bring in the IMPORT menu, which can be helpful to bring in other data tools. In this case, we're bringing in the **import** block.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **sklearn.metrics**, which helps us with the predictions.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the **import** and **package** together in a single variable. This handy feature helps cut down on all the typing later on. You can call it whatever is easiest for you to remember. In the example below, we’ve put everything into **metrics**, and we type it in the open area.



**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!
<details>
    <summary>Click to see the answer...</summary>
    
![](https://pbs.twimg.com/media/GacAgr6W4AAT6At?format=png&name=small
)

</details>

In [ ]:
#blocks code


#### Freehand

**Step 1 - Starting the import**

First, we need to set up a “command” to tell the computer what to do. In this case, “command” is to “import” to bring the add-on package in.



**Step 2 - Telling what library to import**

In the text area, we type the name of the library we want to import. A library is like an extra thing we bring in to give us more coding abilities. In our case, we will type out **sklearn.metrics**, which will provide us with some cool data manipulation features. Some packages have a lengthy name, so we will want to make our own nickname for this.



**Step 3 - Renaming the library so it’s easy to remember**

Once you are done, put the ‘import’ and ‘package’ together in a single variable. This handy feature helps cut down on all the typing later on. Feel free to use whatever name you want that will help you remember it later on. In the example below, we’ve put everything into **metrics**.



**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
    
![](https://pbs.twimg.com/media/GacAi_vXgAAdT_8?format=png&name=small)

</details>


**Your Turn**: Indentation aligned with Step. Bolded and colon. 2-3 sentence description. Please center image


In [16]:
#freehand code 


**Explanation**: *The metrics library provides tools to measure and evaluate the performance of machine learning models. By using `metrics`, we can check how well our model is working*.


### Goal 15: Assessing the Performance of the Predictions on Test Dataset Using R2  

So how well did our predictions do? Let’s calculate R2 to help us think about the performance of predictions on the testing dataset

#### Blockly

**Step 1 - Prepare the R2 score calculation from the metrics library**

To calculate the correctness of the model predictions, we will use the r2_score() function from the metrics library. This correctness score will measure the percentage of correct predictions.  

From the Variables menu, get a DO block for the **metrics** variable. With that, select the **r2_score** operation. This operation will compare the correctness of the model and the test label with the predicted values.



**Step 2 - Calculate the linear model’s correctness**

The r2_score() function takes 2 parameters to calculate the correctness score and helps measure the percentage of correct predictions. So let’s compare **CO2Emission** from the test dataset and **predictions** from the model we just created.

From the Lists menu, get a dictVariable block and select the test variable. From the Text men,u get a Quote “” block to inform the label name **”CO2Emission”**. This list will be used as the true labels for the accuracy calculation.



**Step 3 -  Compare testing labels with the predicted values**

From the Lists menu, get a dictVariable block and select "**test**". Get then a Quote “” block from the Text menu, and type "**CO2Emission**" column.


**Step 4 - Connect the blocks to run the code**

Connect the blocks and run the code!
<details>
    <summary>Click to see the answer...</summary>
    
![](https://pbs.twimg.com/media/GcNQTiBWsAASJbl?format=jpg&name=small)

</details>

In [ ]:
#blocks code


#### Freehand


**Step 1 - Prepare the R2 score calculation from the metrics library**

To calculate the correctness of the model predictions, we will use the **r2_score**() method from the metrics library.  This correctness score will measure the percentage of correct predictions.

`metrics.r2_score()`



**Step 2 - Calculate linear model’s correctness**

The accuracy_score() function takes 2 parameters to calculate the correctness score and help measure the percentage of correct predictions. So let’s compare **CO2Emission** from the test dataset and **predictions** from the model we just created.
The **r2_score**() method takes 2 parameters to calculate the accuracy score.
- Test data labels:** test[‘CO2Emission’]**
- The predicted labels: **predictions**



**Step 3 - Compare testing labels with the predicted values**

So let’s compare CO2Emission from the test dataset and **predictions** from the model we just created.

`metrics.r2_score(test[‘CO2Emission’], predictions)`



**Step 4 - Run the code**

Hit ‘control’ and ‘enter’ at the same time to run the code!

<details>
    <summary>Click to see the answer...</summary>
    
![](https://pbs.twimg.com/media/GcNPLE0XkAETUid?format=jpg&name=small)

</details>

**Your Turn**: Give it a try! Here we use a metric called r2_score() to verify how accurate our model is by comparing our test labels with predictions!


In [ ]:
#freehand code 


**Explanation**: *You have calculated a metric called R-squared to evaluate how well a model’s predictions match the actual values. The R-squared score is a number between 0 and 1. If it’s close to 1, it means the model’s predictions are very close to the actual values, so it’s doing a good job. If it’s closer to 0, it means the predictions aren’t very accurate. In short, a higher R-squared means a better model*.

## WHAT DID YOU LEARN?



In this lesson on multiple linear regression, you learned several important concepts and skills. We started by understanding what a multiple regression model is and how it builds on simple linear regression by adding more predictors. We then explored interaction effects, where the effect of one predictor variable depends on another. You also learned about feature selection, which helps identify the most important variables for predicting an outcome. We discussed categorical variables and how to use dummy variables to include them in our analysis. Throughout the lesson, you got hands-on experience with practical examples, ensuring you not only understood the theory but also gained practical skills in applying multiple linear regression.


## WHAT’S NEXT?

[Logistic Regression](Logistic_Regression.ipynb)

## TELL ME MORE


- [Datawhys Multiple Linear Regression Notebook](https://github.com/memphis-iis/datawhys-content-notebooks-python/blob/master/Multiple-linear-regression.ipynb)
- [Datawhys Multiple Linear Regression Problem-Solving Notebook](https://github.com/memphis-iis/datawhys-content-notebooks-python/blob/master/Multiple-linear-regression-PS.ipynb)
